core/iwasm/fast-jit/fe/jit_emit_memory.c

```diff
@@ -5,18 +5,317 @@
+    if (!jit_emit_exception(cc, EXCE_OUT_OF_BOUNDS_MEMORY_ACCESS, JIT_OP_BGTU,
+                            cc->cmp_reg, 0)) {
+        goto fail;
+    }
+
+    return offset1;
+fail:
+    return 0;
+}
+#endif
+
+static JitReg
+check_and_seek(JitCompContext *cc, JitReg addr, uint32 offset, uint32 bytes)
+{
+    JitReg memory_boundary, offset1, memory_data, maddr;
+    /* the default memory */
+    uint32 mem_idx = 0;
+
+    /* ---------- check ---------- */
+    /* 1. shortcut if the memory size is 0*/
+    if (0 == cc->cur_wasm_module->memories[mem_idx].init_page_count) {
+        JitReg memory_inst, cur_mem_page_count;
+
+        /* if (cur_mem_page_count == 0) goto EXCEPTION */
+        memory_inst = get_memory_inst_reg(cc->jit_frame, mem_idx);
+        cur_mem_page_count = jit_cc_new_reg_I32(cc);
+        GEN_INSN(LDI32, cur_mem_page_count, memory_inst,
+                 NEW_CONST(I32, offsetof(WASMMemoryInstance, cur_page_count)));
+        GEN_INSN(CMP, cc->cmp_reg, cur_mem_page_count, NEW_CONST(I32, 0));
+        if (!jit_emit_exception(cc, EXCE_OUT_OF_BOUNDS_MEMORY_ACCESS,
+                                JIT_OP_BEQ, cc->cmp_reg, 0)) {
+            goto fail;
+        }
+    }
+
+    /* 2. a complete boundary check */
+    memory_boundary = get_memory_boundary(cc, mem_idx, bytes);
+    if (!memory_boundary)
+        goto fail;
+
+#if UINTPTR_MAX == UINT64_MAX
+    offset1 = check_and_seek_on_64bit_platform(cc, addr, NEW_CONST(I64, offset),
+                                               memory_boundary);
+    if (!offset1)
+        goto fail;
+#else
+    offset1 = check_and_seek_on_32bit_platform(cc, addr, NEW_CONST(I32, offset),
+                                               memory_boundary);
+    if (!offset1)
+        goto fail;
+#endif
+
+    /* ---------- seek ---------- */
+    memory_data = get_memory_data_reg(cc->jit_frame, mem_idx);
+    maddr = jit_cc_new_reg_ptr(cc);
+    GEN_INSN(ADD, maddr, memory_data, offset1);
+
+    return maddr;
+fail:
+    return 0;
+}
 
 bool
 jit_compile_op_i32_load(JitCompContext *cc, uint32 align, uint32 offset,
                         uint32 bytes, bool sign, bool atomic)
 {
+    JitReg addr, maddr, value;
+
+    POP_I32(addr);
+
+    maddr = check_and_seek(cc, addr, offset, bytes);
+    if (!maddr) {
+        goto fail;
+    }
+
+    value = jit_cc_new_reg_I32(cc);
+    switch (bytes) {
+        case 1:
+        {
+            if (sign) {
+                JitReg tmp1 = jit_cc_new_reg_I32(cc);
+                JitReg tmp2 = jit_cc_new_reg_I32(cc);
+                GEN_INSN(LDI8, tmp1, maddr, NEW_CONST(I32, 0));
+                GEN_INSN(SHL, tmp2, tmp1, NEW_CONST(I32, 24));
+                GEN_INSN(SHRS, value, tmp2, NEW_CONST(I32, 24));
+            }
+            else {
+                JitReg tmp1 = jit_cc_new_reg_I32(cc);
+                GEN_INSN(LDU8, tmp1, maddr, NEW_CONST(I32, 0));
+                GEN_INSN(AND, value, tmp1, NEW_CONST(I32, 0x000000FF));
```

Comment:
Use `LDU8` is already enough, it will unsigned extend to i32, see:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/dev/fast_jit/core/iwasm/fast-jit/cg/x86-64/jit_codegen_x86_64.cpp#L543